### PR TITLE
[3/9] final_lr_ratio

### PR DESCRIPTION
**Pull request by garrett361**
Alters the SFT linear learning rate scheduler such that the final lr is configurable, rather than having it run down to zero.  The final learning rate is  `learning_rate * final_lr_ratio`

**NOTE**: this changes the default behavior `--final_lr_ratio 0.1`, whereas previously `finetune.py` effectively had `--final_lr_ratio 0.0`.  The 0.1 default is inspired by Qwen.

**NOTE**: only implemented for the `linear` scheduler. This PR disables other scheduler options. 

|   |  PR |                     Title                      |
| - | --- | ---------------------------------------------- |
| 1 | #15 | padding-free                                   |
| 2 | #16 | clean_checkpoints_at_end                       |
| 3 | >17 | final_lr_ratio                                 |
| 4 | #18 | add_seed_and_date_to_run_name                  |
| 5 | #19 | additional_model_arguments                     |
| 6 | #20 | sync_each_batch=True grad acc                  |
| 7 | #21 | no grad acc averaging for sum losses           |
| 8 | #22 | extra reporting                                |
| 9 | #23 | local_main_process_first when building dataset |

### CODE REVIEW

File: open_instruct/finetune.py [hunk 1]
```diff
@@ -373,6 +373,12 @@ class FlatArguments:
             "help": "Whether to clean up all previous checkpoints at the end of the run.",
         },
     )
+    final_lr_ratio: float = field(
```
Comment (fabianlim) [inline open_instruct/finetune.py:376]: default should be `None`, it should be needed unless `linear` scheduler is used right?

File: open_instruct/finetune.py [hunk 2]
```diff
@@ -708,11 +714,24 @@ def main(args: FlatArguments, tc: TokenizerConfig):
     num_training_steps_for_scheduler = (
         args.max_train_steps if overrode_max_train_steps else args.max_train_steps * accelerator.num_processes
     )
+
+    # HACK: @goon - final_lr_ratio assumes a linear scheduler, so putting this assert in to sanity
+    # check.
+    assert args.lr_scheduler_type == "linear"
```
Comment (fabianlim) [inline open_instruct/finetune.py:720]: this kind of check should be in the `__post__init__` of the `FlattenedArgs`

File: open_instruct/finetune.py [hunk 3]
```diff
@@ -708,11 +714,24 @@ def main(args: FlatArguments, tc: TokenizerConfig):
...
+    assert 1.0 > args.final_lr_ratio >= 0.0
+
+    # HACK: @goon - adjust num_training_steps_for_scheduler so that the final LR is learning_rate *
```
Comment (fabianlim) [inline open_instruct/finetune.py:723]: we need to have the original behavior if `final_lr_ratio` is None

**Review by fabianlim — CHANGES_REQUESTED** [review]
final_lr_ratio should not be a required setting, if scheduler is not linear it should have have any effect

File: open_instruct/finetune.py [hunk 4]
```diff
@@ -708,11 +720,19 @@ def main(args: FlatArguments, tc: TokenizerConfig):
     num_training_steps_for_scheduler = (
         args.max_train_steps if overrode_max_train_steps else args.max_train_steps * accelerator.num_processes
     )
+
+    num_warmup_steps = int(num_training_steps_for_scheduler * args.warmup_ratio)
+    if args.final_lr_ratio is not None and args.lr_scheduler_type == "linear":
+        # Correct num_training_steps_for_scheduler to respect final_lr_ratio for a linear scheduler
+        num_training_steps_for_scheduler = (
```
[the remark this answers is not in the document]
Comment (garrett361) [reply open_instruct/finetune.py:727]: Correct, it's just defined here, and maybe updated if the user specifies `final_lr_ratio` and is using a `linear` scheduler.
